lib/Dialect/Pipeline/PipelineOps.cpp

```diff
@@ -327,6 +327,154 @@ LogicalResult ReturnOp::verify() {
+    return;
+
+  p << "regs(";
+  size_t clockGateStartIdx = 0;
+  llvm::interleaveComma(
+      llvm::zip(registers, registerTypes, clockGatesPerRegister), p,
+      [&](auto it) {
+        auto &[reg, type, nClockGatesAttr] = it;
+        p << reg << " : " << type;
+        int64_t nClockGates =
+            nClockGatesAttr.template cast<IntegerAttr>().getInt();
+        if (nClockGates == 0)
+          return;
+        p << " gated by [";
+        llvm::interleaveComma(clockGates.slice(clockGateStartIdx, nClockGates),
+                              p);
+        p << "]";
+        clockGateStartIdx += nClockGates;
+      });
+  p << ")";
+}
+
+void StageOp::build(OpBuilder &odsBuilder, OperationState &odsState,
+                    Block *dest, ValueRange registers,
+                    ValueRange passthroughs) {
+  odsState.addSuccessors(dest);
+  odsState.addOperands(registers);
+  odsState.addOperands(passthroughs);
+  odsState.addAttribute("operand_segment_sizes",
+                        odsBuilder.getDenseI32ArrayAttr(
+                            {static_cast<int32_t>(registers.size()),
+                             static_cast<int32_t>(passthroughs.size()),
+                             /*clock gates*/ static_cast<int32_t>(0)}));
+  llvm::SmallVector<int64_t> clockGatesPerRegister(registers.size(), 0);
+  odsState.addAttribute("clockGatesPerRegister",
+                        odsBuilder.getI64ArrayAttr(clockGatesPerRegister));
+}
+
+void StageOp::build(
+    OpBuilder &odsBuilder, OperationState &odsState, Block *dest,
+    ValueRange registers, ValueRange passthroughs,
+    const llvm::DenseMap<Value, llvm::SmallVector<Value>> &clockGates) {
+  odsState.addSuccessors(dest);
+  odsState.addOperands(registers);
+  odsState.addOperands(passthroughs);
+
+  // Get a sorted list of clock gates wrt. the registers.
+  llvm::SmallVector<int64_t> clockGatesPerRegister;
+  llvm::SmallVector<Value> sortedClockGates;
+  for (Value reg : registers) {
+    size_t nClockGates = 0;
+    if (auto it = clockGates.find(reg); it != clockGates.end()) {
+      llvm::append_range(sortedClockGates, it->second);
+      nClockGates = it->second.size();
+    }
+    clockGatesPerRegister.push_back(nClockGates);
+  }
+  odsState.addOperands(sortedClockGates);
+  odsState.addAttribute(
+      "operand_segment_sizes",
+      odsBuilder.getDenseI32ArrayAttr(
+          {static_cast<int32_t>(registers.size()),
+           static_cast<int32_t>(passthroughs.size()),
+           /*clock gates*/ static_cast<int32_t>(sortedClockGates.size())}));
+  odsState.addAttribute("clockGatesPerRegister",
+                        odsBuilder.getI64ArrayAttr(clockGatesPerRegister));
+}
+
+ValueRange StageOp::getClockGatesForReg(unsigned regIdx) {
+  assert(regIdx < getRegisters().size() && "register index out of bounds.");
+
+  // This could be optimized quite a bit if we didn't store clock gates per
+  // register as an array of sizes... TODO: look into using properties and maybe
```

Comment:
uber nit: Some code editors recognize `// TODO:` and highlight it appropriately. Your format here may not be picked up by some editors.